### PR TITLE
Add method for extending the Stackdriver JSON Layout

### PR DESCRIPTION
**Pull request by dzou**
This adds a method for extending the `StackdriverJsonLayout` so that it is more customizable. This is modeled after the logging client library's method of using `loggingEventEnhancer` elements in the logback.xml to enhance log entries (see [javadoc](https://github.com/googleapis/java-logging-logback/blob/master/src/main/java/com/google/cloud/logging/logback/LoggingAppender.java#L74)).

* Added way to enhance log entries via `loggingEventEnhancer` element:
 
  ```
    <appender name="CONSOLE_JSON" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
            <layout class="com.google.cloud.spring.logging.StackdriverJsonLayout">
                <loggingEventEnhancer>
                    com.google.cloud.spring.logging.extensions.LogstashLoggingEventEnhancer
                </loggingEventEnhancer>
            </layout>
        </encoder>
    </appender>
  ```

* Added `extensions` package in logging module to be the home of future logging enhancers/extensions. I added one for the Logstash markers that the user requested #198 which requires a `<provided>` scope dependency on Logstash libraries.

Fixes https://github.com/GoogleCloudPlatform/spring-cloud-gcp/issues/198

### CODE REVIEW

**Review by ttomsu — APPROVED** [review]

**Review by elefeint — CHANGES_REQUESTED** [review]
Can you add a blurb to the refdoc about how end users would enable these? Similar to what you have in javadoc.

File: spring-cloud-gcp-logging/pom.xml [hunk 1]
```diff
@@ -26,6 +26,14 @@
 			<artifactId>google-cloud-logging-logback</artifactId>
 		</dependency>
 
+		<!-- Logging Extensions -->
+		<dependency>
+			<groupId>net.logstash.logback</groupId>
+			<artifactId>logstash-logback-encoder</artifactId>
+			<version>6.6</version>
+            <scope>provided</scope>
```
[the remark this answers is not in the document]
Comment (dzou) [reply spring-cloud-gcp-logging/pom.xml]: Done.

File: spring-cloud-gcp-logging/pom.xml [hunk 2]
```diff
@@ -26,6 +26,14 @@
 			<artifactId>google-cloud-logging-logback</artifactId>
 		</dependency>
 
+		<!-- Logging Extensions -->
+		<dependency>
+			<groupId>net.logstash.logback</groupId>
+			<artifactId>logstash-logback-encoder</artifactId>
+			<version>6.6</version>
```
[the remark this answers is not in the document]
Comment (dzou) [reply spring-cloud-gcp-logging/pom.xml]: Done.

File: spring-cloud-gcp-logging/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dzou) [reply spring-cloud-gcp-logging/pom.xml]: Done.

**Comment by dzou** [comment]
> Can you add a blurb to the refdoc about how end users would enable these? Similar to what you have in javadoc.

I ended up adding an entry to the table and then adding to the example logback.xml in the docs: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/d98ed5ceb2abc42c0833e35b3884cebda260a22a/docs/src/main/asciidoc/logging.adoc#log-via-console

I considered adding a paragraph to describe it more but decided against it because I felt it would bloat the docs, and users can figure it out by following the link to the examples and seeing the example logback.

File: docs/src/main/asciidoc/logging.adoc [hunk 3]
```diff
@@ -146,6 +146,7 @@ This setting is only evaluated if `includeFormattedMessage` is `true`
 | `includeException` | `false` | Should the stacktrace be included as a own field
 | `serviceContext` | none | Define the Stackdriver service context data (service and version). This allows filtering of error reports for service and version in the https://console.cloud.google.com/errors[Google Cloud Error Reporting View].
 | `customJson` | none | Defines custom json data. Data will be added to the json output.
+| `loggingEventEnhancer` | none | Class names implementing `JsonLoggingEventEnhancer` which modifies the JSON logging output (link:https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/master/spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/extensions[examples]).
```
Comment (elefeint) [inline docs/src/main/asciidoc/logging.adoc:149]: I'd probably mention that a Logstash tag-enriching implementation is available in the extensions package, just to make it discoverable through search engines.

Comment (elefeint) [inline docs/src/main/asciidoc/logging.adoc:149]:
```suggestion
| `loggingEventEnhancer` | none | Name of a class implementing `JsonLoggingEventEnhancer` which modifies the JSON logging output (link:https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/master/spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/extensions[examples]).
```

Comment (elefeint) [reply docs/src/main/asciidoc/logging.adoc:149]: Also, if the tag is repeatable, mention that. It's a repeatable XML tag, right? -- it's not like you can specify multiple classes in the same tag.

**Review by elefeint — APPROVED** [review]
I've been on a grammar nitpicking binge. Hypocritical, really, since I am not a native speaker.

File: spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java [hunk 4]
```diff
@@ -157,6 +162,20 @@ public void setCustomJson(String json) {
...
+	 * Add additional logging enhancers that implement {@link JsonLoggingEventEnhancer}.
+	 * @param enhancerClassName class name of the layout enhancer
+	 */
+	public void addLoggingEventEnhancer(String enhancerClassName) {
+		try {
+			Class<JsonLoggingEventEnhancer> clz =
+					(Class<JsonLoggingEventEnhancer>) Loader.loadClass(enhancerClassName.trim());
+			loggingEventEnhancers.add(clz.getDeclaredConstructor().newInstance());
+		}
+		catch (Exception ex) {
```
Comment (meltsufin) [inline spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java:175]: Is this for `ClassNotFoundException`?

**Review by meltsufin — APPROVED** [review]
Should we mention the existence of `LogstashLoggingEventEnhancer` in the refdoc at all? Maybe just in the property table?

File: spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java (hunk 4 above)
Comment (dzou) [reply spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java:175]: Yeah there's a couple checked exceptions -- ClassNotFound, MethodNotFound (for invoking the constructor), and a few others.

File: docs/src/main/asciidoc/logging.adoc (hunk 3 above)
Comment (dzou) [reply docs/src/main/asciidoc/logging.adoc]: Done and Done.

**Comment by dzou** [comment]
> Should we mention the existence of `LogstashLoggingEventEnhancer` in the refdoc at all? Maybe just in the property table?

Done.

**Comment by ttomsu** [comment]
You'll have to get this class's code coverage up to get a green Sonar build: https://sonarcloud.io/component_measures?id=GoogleCloudPlatform_spring-cloud-gcp&metric=new_coverage&pullRequest=208&selected=GoogleCloudPlatform_spring-cloud-gcp%3Aspring-cloud-gcp-logging%2Fsrc%2Fmain%2Fjava%2Fcom%2Fgoogle%2Fcloud%2Fspring%2Flogging%2Fextensions%2FLogstashLoggingEventEnhancer.java&view=list

**Review by meltsufin — APPROVED** [review]

**Review by elefeint — APPROVED** [review]

**Pull request merged**